mplc/mpl_utils.py

```diff
@@ -0,0 +1,177 @@
+        plt.plot(self.history['mpl_model']['val_loss'][:self.mpl.epoch_index + 1, self.mpl.minibatch_count])
+        plt.ylabel("Loss")
+        plt.xlabel("Epoch")
+        plt.savefig(self.save_folder / "graphs/federated_training_loss.png")
+        plt.close()
+
+        plt.figure()
+        plt.plot(self.history['mpl_model']['val_accuracy'][:self.mpl.epoch_index + 1, self.mpl.minibatch_count])
+        plt.ylabel("Accuracy")
+        plt.xlabel("Epoch")
+        plt.ylim([0, 1])
+        plt.savefig(self.save_folder / "graphs/federated_training_acc.png")
+        plt.close()
+
+        plt.figure()
+        for key, value in self.history.items():
+            plt.plot(value['val_accuracy'][:self.mpl.epoch_index + 1, self.mpl.minibatch_count],
+                     label=(f'partner {key}' if key != 'mpl_model' else key))
+        plt.title("Model accuracy")
+        plt.ylabel("Accuracy")
+        plt.xlabel("Epoch")
+        plt.legend()
+        plt.ylim([0, 1])
+        plt.savefig(self.save_folder / "graphs/all_partners.png")
+        plt.close()
+
+
+class Aggregator(ABC):
+    def __init__(self, mpl):
+        """
+        :type mpl: MultiPartnerLearning
+        """
+        self.mpl = mpl
+        self.aggregation_weights = np.zeros(self.mpl.partners_count)
+
+    def aggregate_model_weights(self):
+        """Aggregate model weights from the list of partner's models, with a weighted average"""
+
+        weights_per_layer = list(zip(*[partner.model_weights for partner in self.mpl.partners_list]))
+        new_weights = list()
+
+        for weights_for_layer in weights_per_layer:
+            avg_weights_for_layer = np.average(
+                np.array(weights_for_layer), axis=0, weights=self.aggregation_weights
+            )
+            new_weights.append(list(avg_weights_for_layer))
+
+        return new_weights
+
+
+class UniformAggregator(Aggregator):
+    def __init__(self, mpl):
+        super(UniformAggregator, self).__init__(mpl)
+        self.aggregation_weights = [1 / self.mpl.partners_count] * self.mpl.partners_count
+
+
+class DatavolumeAggregator(Aggregator):
+    def __init__(self, mpl):
+        super(DatavolumeAggregator, self).__init__(mpl)
+        partners_sizes = [partner.data_volume for partner in self.mpl.partners_list]
+        self.aggregation_weights = partners_sizes / np.sum(partners_sizes)
+
+
+class ScoresAggregator(Aggregator):
+    def __init__(self, mpl):
+        super(ScoresAggregator, self).__init__(mpl)
+
+    def prepare_aggregation_weights(self):
+        last_scores = [partner.last_round_score for partner in self.mpl.partners_list]
+        self.aggregation_weights = last_scores / np.sum(last_scores)
+
+    def aggregate_model_weights(self):
+        self.prepare_aggregation_weights()
+        super(ScoresAggregator, self).aggregate_model_weights()
+
+
+# Supported aggregation weights approaches
+AGGREGATORS = {
+    "uniform": UniformAggregator,
+    "data-volume": DatavolumeAggregator,
```

Comment:
change main config file accordinly

Reply:
It is not very clear for me what I could change in the main.py file. Because from a scenario's point of view nothing change, and main.py only deals with scenario.